### PR TITLE
[core] Add GeoIp info to security logs

### PR DESCRIPTION
**Pull request by chalasr**
| Q             | A
| ------------- | --- 
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Fixed issues | -

To enable collecting geoip info for a given platform, set the `geoip.maxmind_license_key` platform option to a valid [maxmind geoip2](https://dev.maxmind.com/geoip/geoip2/geolite2/) license key and run `composer install` (or `composer run-script setup-geoip`).
A priori, we're going to use the same key for all our platforms though.

On composer install/update, the database is downloaded (and overwritten if it already exists). That's the only time where the license key is used.
In order to not be bothered by rate limiting while working on the project, I did not add the key to our CI. 

Then, each time the platform writes to the security logs table, the local geoip database is hit to enrich the log entry.
The runtime overhead is not significant and anyway, we are in the process of making our functional logging asynchronous.

### CODE REVIEW

**Comment by chalasr** [comment]
Ready

File: src/main/core/GeoIp/Database/MaxMindGeoIpDatabaseDownloader.php [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+
+use Psr\Log\LoggerInterface;
+use Psr\Log\NullLogger;
+use Symfony\Component\Filesystem\Exception\IOException;
+use Symfony\Component\Filesystem\Filesystem;
+use Symfony\Component\HttpClient\HttpClient;
+use Symfony\Contracts\HttpClient\Exception\ExceptionInterface;
+use Symfony\Contracts\HttpClient\HttpClientInterface;
+
+class MaxMindGeoIpDatabaseDownloader
```
Comment (Elorfin) [inline src/main/core/GeoIp/Database/MaxMindGeoIpDatabaseDownloader.php:13]: implements `LoggerAwareInterface`

File: src/main/core/GeoIp/Database/MaxmindGeoIpScriptHandler.php [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+use Claroline\BundleRecorder\Logger\ConsoleIoLogger;
+use Composer\Script\Event;
+
+class MaxmindGeoIpScriptHandler
+{
+    public static function refreshDatabase(Event $event): void
+    {
+        $logger = new ConsoleIoLogger($event->getIo());
+        $geoIpLicenseKey = getenv('MAXMIND_GEOIP_LICENSE_KEY');
```
Comment (Elorfin) [inline src/main/core/GeoIp/Database/MaxmindGeoIpScriptHandler.php:13]: You should reuse the `platform_options` param you've declared.

FMI : Do we need a different key for each platform ?

File: src/main/core/GeoIp/GeoIpInfoProviderInterface.php [hunk 3]
```diff
@@ -0,0 +1,8 @@
+<?php
+
+namespace Claroline\CoreBundle\GeoIp;
+
+interface GeoIpInfoProviderInterface
+{
+    public function getGeoIpInfo(string $ip): ?GeoIpinfo;
```
Comment (Elorfin) [inline src/main/core/GeoIp/GeoIpInfoProviderInterface.php:7]:
```suggestion
    public function getGeoIpInfo(string $ip): ?GeoIpInfo;
```

File: src/main/core/Library/Configuration/PlatformDefaults.php [hunk 4]
```diff
@@ -206,6 +206,9 @@ public function getDefaultParameters()
                 'enabled' => false,
                 'currency' => 'euro',
             ],
+            'maxmind_geoip' => [
+                'license_key' => null,
```
Comment (Elorfin) [inline src/main/core/Library/Configuration/PlatformDefaults.php:210]: It is not used.

**Review by Elorfin — CHANGES_REQUESTED** [review]
We can discuss about it if you want.

**Comment by Elorfin** [comment]
Another thing : namespace `Claroline\CoreBundle\GeoIp` seems too high in the hierarchy for such a small feature. It's not ideal but I would have hidden it in the `Library` (or elsewhere if you have a better idea).

File: src/main/core/GeoIp/Database/MaxMindGeoIpDatabaseDownloader.php [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+    private $filesystem;
+    private $tempDir;
+
+    public function __construct(string $licenseKey, ?LoggerInterface $logger = null, ?HttpClientInterface $httpClient = null, ?Filesystem $filesystem = null, ?string $tempDir = null)
+    {
+        $this->licenseKey = $licenseKey;
+        $this->logger = $logger ?? new NullLogger();
+        $this->httpClient = $httpClient ?? HttpClient::create();
+        $this->filesystem = $filesystem ?? new Filesystem();
+        $this->tempDir = $tempDir ?? sys_get_temp_dir();
```
[the remark this answers is not in the document]
Comment (chalasr) [reply src/main/core/GeoIp/Database/MaxMindGeoIpDatabaseDownloader.php:30]: > Also note there is a platform_options param tmp_dir/server.tmp_dir for the last one.

Changing for a console command will allow getting the platform config injected 👍

**Comment by chalasr** [comment]
@Elorfin comments addressed

File: src/main/core/Library/GeoIp/Database/MaxMindGeoIpDatabaseDownloader.php [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+        if (200 !== $statusCode) {
+            $this->logger->warning('GeoIp database file could not be downloaded.', ['status_code' => $statusCode, 'error' => $response->getInfo()['error'] ?? null]);
+
+            return;
+        }
+
+        // Download and dump the archive as a temp file (chunk-by-chunk)
+        $tempDir = "$this->tempDir/claroline-geoip";
+        $this->filesystem->mkdir($tempDir);
+        $tmpArchive = $this->filesystem->tempnam(sys_get_temp_dir().'/claroline-geoip', 'maxmind-geoip').'.tar.gz';
```
Comment (Elorfin) [inline src/main/core/Library/GeoIp/Database/MaxMindGeoIpDatabaseDownloader.php:48]: You should reuse `$tempDir`

File: src/main/core/Library/GeoIp/Database/MaxMindGeoIpDatabaseDownloader.php [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+        }
+
+        // Move the database file to the destination (overwrite if exists)
+        $databaseDirectory = basename($archive->current()->getPathname());
+        $this->filesystem->rename("$extractDirectory/$databaseDirectory/".self::DATABASE_FILENAME, $destinationDir.'/'.self::DATABASE_FILENAME, true);
+
+        $this->logger->info('GeoIp database successfully downloaded.');
+
+        // Cleanup filesystem
+        $this->filesystem->remove("$this->tempDir/claroline-geoip");
```
Comment (Elorfin) [inline src/main/core/Library/GeoIp/Database/MaxMindGeoIpDatabaseDownloader.php:85]: You should reuse `$tempDir`

File: src/main/core/Resources/config/services/library.yml [hunk 8]
```diff
@@ -1,3 +1,7 @@
+parameters:
+    claroline.param.geoip_db_path: '%kernel.project_dir%/var/geoip/GeoLite2-City.mmdb'
+    env(MAXMIND_GEOIP_LICENSE_KEY): ''
```
Comment (Elorfin) [inline src/main/core/Resources/config/services/library.yml:3]: move it in `src/main/core/Resources/config/parameters.yml`.

Comment (Elorfin) [reply src/main/core/Resources/config/services/library.yml:3]: We may want a `platform_options` param for `MAXMIND_GEOIP_LICENSE_KEY` instead.

Comment (chalasr) [reply src/main/core/Resources/config/services/library.yml]: Option added

**Comment by chalasr** [comment]
Ready for another round.

**Review by Elorfin — APPROVED** [review]
👍

**Pull request merged**